[test_windowed_sdpa.py:107] | Sanity tests nightly debug run with LLK asserts | Sanity tests nightly debug run (Ubuntu 22.04 with LLK asserts) / ttnn-sanity-tests / ttnn sdpa group [wh_n300_civ2]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] test_windowed_sdpa_smoke hangs with device timeout (TT_THROW system_memory_manager.cpp:780) on wh_n300_civ2 LLK-asserts sanity run
<!-- AUTO-TRIAGE-SUMMARY-END -->
**Workflow:** Sanity tests nightly debug run with LLK asserts
**Job:** Sanity tests nightly debug run (Ubuntu 22.04 with LLK asserts) / ttnn-sanity-tests / ttnn sdpa group [wh_n300_civ2]

### Regression timeline
- Last passing commit: N/A
- First failing commit: `66d22c1f3a9d` on 2026-07-28T03:02:15Z — https://github.com/tenstorrent/tt-metal/actions/runs/30324799136/job/90303792780
- Most recent failing commit: `6935d7117e73` on 2026-07-29T03:01:58Z — https://github.com/tenstorrent/tt-metal/actions/runs/30418433000/job/90483694556

### Failing job URLs (last 2 runs)
- https://github.com/tenstorrent/tt-metal/actions/runs/30418433000/job/90483694556
- https://github.com/tenstorrent/tt-metal/actions/runs/30324799136/job/90303792780

### Failing test path(s)
- `tests/ttnn/unit_tests/operations/sdpa/test_windowed_sdpa.py::test_windowed_sdpa_smoke[fp32acc-bf16-num_heads=1-s128_w2]`

### Error excerpt (terminal failure from logs)
```
=================================== FAILURES ===================================
__________ test_windowed_sdpa_smoke[fp32acc-bf16-num_heads=1-s128_w2] __________
...
>       out = ttnn.to_torch(out_tt).to(torch.float32)
tests/ttnn/unit_tests/operations/sdpa/test_windowed_sdpa.py:107:
...
>               result = self.function(*function_args, **function_kwargs)
E               RuntimeError: TT_THROW @ /project/tt_metal/impl/dispatch/system_memory_manager.cpp:780: tt::exception
E               info:
E               TIMEOUT: device timeout, potential hang detected, the device is unrecoverable

=========================== short test summary info ============================
FAILED tests/ttnn/unit_tests/operations/sdpa/test_windowed_sdpa.py::test_windowed_sdpa_smoke[fp32acc-bf16-num_heads=1-s128_w2] - RuntimeError: TT_THROW @ /project/tt_metal/impl/dispatch/system_memory_manager.cpp:780: tt::exception
======= 1 failed, 56 passed, 50 skipped, 1 warning in 202.89s (0:03:22) ========
```

### Reproduction steps
1. Build/pull the ttnn wheel matching the failing commit (e.g. `ttnn-0.65.1rc17.dev7421+g6935d7117e-cp310-cp310-manylinux_2_34_x86_64.whl`) on a wh_n300 (civ2) machine.
2. Set `TT_METAL_LLK_ASSERTS=1` and `TT_METAL_OPERATION_TIMEOUT_SECONDS=5` as used in the CI job.
3. Run: `pytest tests/ttnn/unit_tests/operations/sdpa/test_windowed_sdpa.py::test_windowed_sdpa_smoke -k "fp32acc and bf16 and num_heads=1 and s128_w2"`
4. Observe the device hang/timeout and `TT_THROW` at `system_memory_manager.cpp:780`.

### Notes
- This is a device hang detected via the dispatch timeout mechanism (`TT_METAL_OPERATION_TIMEOUT_SECONDS=5`), not a numerical/PCC assertion failure — the op never completes and the device becomes unrecoverable.
- Determinism check:
  - Run 1 (commit `6935d7117e73`, 2026-07-29): `FAILED tests/ttnn/unit_tests/operations/sdpa/test_windowed_sdpa.py::test_windowed_sdpa_smoke[fp32acc-bf16-num_heads=1-s128_w2] - RuntimeError: TT_THROW @ /project/tt_metal/impl/dispatch/system_memory_manager.cpp:780: tt::exception` with `TIMEOUT: device timeout, potential hang detected, the device is unrecoverable`; summary `1 failed, 56 passed, 50 skipped, 1 warning in 202.89s`.
  - Run 2 (commit `66d22c1f3a9d`, 2026-07-28): identical `FAILED ... test_windowed_sdpa_smoke[fp32acc-bf16-num_heads=1-s128_w2] - RuntimeError: TT_THROW @ /project/tt_metal/impl/dispatch/system_memory_manager.cpp:780: tt::exception` with the same `TIMEOUT` message; summary `1 failed, 56 passed, 50 skipped, 1 warning in 199.32s`.

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: Sanity tests nightly debug run with LLK asserts`
`Auto-triage-job-name: Sanity tests nightly debug run (Ubuntu 22.04 with LLK asserts) / ttnn-sanity-tests / ttnn sdpa group [wh_n300_civ2]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

--- a/tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py
+++ b/tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py
@@ -546,7 +546,12 @@ class RingJointSDPARuntime:
 
 
 def open_ring_joint_sdpa_runtime(
-    mesh_config, *, fp32_dest_acc_en: bool = False, trace_region_size: int = 0, topology: Topology = None
+    mesh_config,
+    *,
+    fp32_dest_acc_en: bool = False,
+    trace_region_size: int = 0,
+    topology: Topology = None,
+    reserve_llk_kernel_config: bool = True,
 ):
     use_ring = mesh_config.sp_size > 2 if topology is None else topology == Topology.Ring
     fabric_config = ttnn.FabricConfig.FABRIC_1D_RING if use_ring else ttnn.FabricConfig.FABRIC_1D
@@ -574,7 +579,7 @@ def open_ring_joint_sdpa_runtime(
         # trace_region_size defaults to 0 (no trace region), leaving every existing caller unchanged; only
         # the trace-replay test asks for one.
         mesh_device_kwargs = {"mesh_shape": mesh_shape, "trace_region_size": trace_region_size}
-        if os.getenv("TT_METAL_LLK_ASSERTS"):
+        if reserve_llk_kernel_config and os.getenv("TT_METAL_LLK_ASSERTS"):
             mesh_device_kwargs["worker_l1_size"] = LLK_ASSERT_WORKER_L1_SIZE
         mesh_device = ttnn.open_mesh_device(**mesh_device_kwargs)
 
@@ -1621,6 +1626,7 @@ def run_ring_joint_sdpa_chunked(
     sliding_window_size: int = None,
     use_attention_sink: bool = False,
     runtime: RingJointSDPARuntime = None,
+    reserve_llk_kernel_config: bool = True,
 ):
     """
     Validate ring joint SDPA chunked-prefill, or verify deterministic replay.
@@ -1705,7 +1711,11 @@ def run_ring_joint_sdpa_chunked(
 
     owns_runtime = runtime is None
     if runtime is None:
-        runtime = open_ring_joint_sdpa_runtime(mesh_config, fp32_dest_acc_en=fp32_dest_acc_en)
+        runtime = open_ring_joint_sdpa_runtime(
+            mesh_config,
+            fp32_dest_acc_en=fp32_dest_acc_en,
+            reserve_llk_kernel_config=reserve_llk_kernel_config,
+        )
     mesh_device = runtime.mesh_device
     topology = runtime.topology
     sp_axis = runtime.sp_axis
@@ -5324,6 +5334,7 @@ def test_ring_mla_chunked_accuracy(model_name, qk_configs, chunk_size):
         qk_configs=qk_configs,
         persistent_buffer_mode="reuse_max",
         use_ring_mla=True,
+        reserve_llk_kernel_config=False,
     )
 
 
@@ -5416,6 +5427,7 @@ def test_ring_mla_chunked_determinism(model_name, qk_configs, chunk_size):
         qk_configs=qk_configs,
         num_iterations=3,
         use_ring_mla=True,
+        reserve_llk_kernel_config=False,
     )
 
 

--- a/tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py
+++ b/tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py
@@ -59,6 +59,10 @@ MESH_CONFIG = MeshConfig.detect()
 
 BATCH_SIZE = 1
 
+# Experimental LLK-assert override: shrink user-allocatable worker L1 by 16 KiB so the
+# kernel-config staging region grows from 69 KiB to 85 KiB on Blackhole.
+LLK_ASSERT_WORKER_L1_SIZE = 1_444_992
+
 
 @dataclass(frozen=True)
 class GPTOSSRingSinkConfig:
@@ -569,7 +573,10 @@ def open_ring_joint_sdpa_runtime(
         mesh_shape = ttnn.MeshShape(mesh_config.tp_size, mesh_config.sp_size)
         # trace_region_size defaults to 0 (no trace region), leaving every existing caller unchanged; only
         # the trace-replay test asks for one.
-        mesh_device = ttnn.open_mesh_device(mesh_shape=mesh_shape, trace_region_size=trace_region_size)
+        mesh_device_kwargs = {"mesh_shape": mesh_shape, "trace_region_size": trace_region_size}
+        if os.getenv("TT_METAL_LLK_ASSERTS"):
+            mesh_device_kwargs["worker_l1_size"] = LLK_ASSERT_WORKER_L1_SIZE
+        mesh_device = ttnn.open_mesh_device(**mesh_device_kwargs)
 
         full_compute_grid = mesh_device.compute_with_storage_grid_size()
         ccl_sub_device_crs = ttnn.CoreRangeSet(

--- a/ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_common.hpp
+++ b/ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_common.hpp
@@ -262,9 +262,9 @@ void recip_block_inplace(uint32_t in_cb, uint32_t num_tiles) {
     CircularBuffer cb_in(in_cb);
     // Precondition: in_cb has num_tiles produced
     // Postcondition: in_cb has num_tiles produced
+    reconfig_data_format_srca(in_cb);
     copy_tile_to_dst_init_short(in_cb);
     recip_tile_init();
-    reconfig_data_format_srca(in_cb);
     pack_reconfig_data_format(in_cb);
 
     cb_in.wait_front(num_tiles);
@@ -300,8 +300,11 @@ void sub_exp_block_bcast_cols_inplace(uint32_t in1_cb, uint32_t reduce_cb, uint3
     // Precondition: in1_cb has rows produced
     // Postcondition: in0_cb has rows*cols produced
     // Postcondition: in1_cb has rows produced
-    sub_bcast_cols_init_short(in0_cb, in1_cb);
+    // llk_unpack_AB_init (inside sub_bcast_cols_init_short) validates the live
+    // unpacker configuration.  Reconfigure first: qk_im can be FP32 while the
+    // row maximum is BF16, notably after applying a windowed BF16 mask.
     reconfig_data_format(in0_cb, in1_cb);
+    sub_bcast_cols_init_short(in0_cb, in1_cb);
 
     // The exponential function uses InputClamping::None for better performance. This version
     // produces incorrect outputs for inputs <~ -88, but those outputs are guaranteed to be negative.
@@ -488,8 +491,8 @@ void mul_block_bcast_cols_inplace(uint32_t in0_cb, uint32_t in1_cb) {
     constexpr uint32_t granularity = cols;
 #endif
 
-    mul_bcast_cols_init_short(in0_cb, in1_cb);
     reconfig_data_format(in0_cb, in1_cb);
+    mul_bcast_cols_init_short(in0_cb, in1_cb);
     pack_reconfig_data_format(in0_cb);
     cb_in0.wait_front(num_tiles);
     cb_in1.wait_front(rows);
@@ -598,8 +601,8 @@ void mul_tiles_bcast_cols_inplace(uint32_t in0_cb, uint32_t in1_cb, uint32_t num
     // Postcondition: in0_cb has num_tiles produced
     // Postcondition: in1_cb has num_tiles produced
 
-    mul_bcast_cols_init_short(in0_cb, in1_cb);
     reconfig_data_format(in0_cb, in1_cb);
+    mul_bcast_cols_init_short(in0_cb, in1_cb);
     pack_reconfig_data_format(in0_cb);
     cb_in0.wait_front(num_tiles);
     cb_in1.wait_front(num_tiles);
@@ -1085,13 +1088,15 @@ void matmul_reduce(uint32_t in1_cb, const uint32_t& out_cb) {
      * Use matmul on Mx1 input to reduce rows within tile to produce Mx1 output.
      */
 
+    // matmul_block_init validates the live reverse-order unpacker setup
+    // (in1_cb -> SrcA, out_cb -> SrcB), so establish it before init.
+    reconfig_data_format(in1_cb, out_cb);
     matmul_block_init(
         out_cb, in1_cb, 0 /*transpose*/, subblock_w /*ct_dim*/, subblock_h /*rt_dim*/, in0_block_w /*kt_dim*/);
 
     constexpr uint32_t output_num_tiles = M * N;
     constexpr uint32_t out_subblock_num_tiles = subblock_h * subblock_w;
 
-    reconfig_data_format(in1_cb, out_cb);
     pack_reconfig_data_format(out_cb);
     cb_in1.wait_front(N);
     cb_out.wait_front(M);

